File: src/commands/doctor.ts
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import pc from "picocolors";
import { loadConfig, validateConfig } from "../config.js";
import { getConfigPath } from "../paths.js";
import { getCredentialStatus } from "../auth.js";
import { findBinary } from "./launch.js";
import { sectionHeader } from "../display.js";

// ── Output Helpers ──────────────────────────────────

function check(msg: string): void {
  console.log(`  ${pc.green("\u2714")} ${msg}`);
}

function cross(msg: string): void {
  console.log(`  ${pc.red("\u2716")} ${msg}`);
}

function warning(msg: string): void {
  console.log(`  ${pc.yellow("\u26A0")} ${msg}`);
}

// ── Diagnostic Checks ───────────────────────────────

function checkConfigFile(): boolean {
  const configPath = getConfigPath();
  if (!fs.existsSync(configPath)) {
    cross(`Config file not found at ${configPath}`);
    return false;
  }

  let raw: string;
  try {
    raw = fs.readFileSync(configPath, "utf-8");
  } catch {
    cross(`Config file unreadable at ${configPath}`);
    return false;
  }

  let parsed: unknown;
  try {
    parsed = JSON.parse(raw);
  } catch {
    cross("Config file contains malformed JSON");
    return false;
  }

  if (!validateConfig(parsed)) {
    cross("Config file has invalid structure");
    return false;
  }

  check("Config file valid");
  return true;
}

function checkPath(): void {
  const localBin = path.join(os.homedir(), ".local", "bin");
  const envPath = process.env["PATH"] ?? "";
  const separator = process.platform === "win32" ? ";" : ":";
  const dirs = envPath.split(separator);

  if (dirs.some((d) => path.normalize(d) === path.normalize(localBin))) {
    check("PATH includes ~/.local/bin");
  } else {
    warning("~/.local/bin is not on PATH");
  }
}

type ShellType = "bash" | "zsh" | "fish" | "powershell";

function detectShell(): ShellType {
  if (process.platform === "win32") {
    return "powershell";
  }
  const shellEnv = process.env["SHELL"] ?? "";
  const base = path.basename(shellEnv);
  if (base === "zsh") return "zsh";
  if (base === "fish") return "fish";
  return "bash";
}

function getShellProfilePath(shell: ShellType): string | null {
  const home = os.homedir();
  switch (shell) {
    case "powershell":
      return path.join(home, "Documents", "PowerShell", "Microsoft.PowerShell_profile.ps1");
    case "bash":
      return path.join(home, ".bashrc");
    case "zsh":
      return path.join(home, ".zshrc");
    case "fish":
      return path.join(home, ".config", "fish", "config.fish");
  }
}

function checkShellIntegration(): void {
  const shell = detectShell();
  const profilePath = getShellProfilePath(shell);

  if (!profilePath) {
    warning(`Could not determine shell profile path for ${shell}`);
    return;
  }

  if (!fs.existsSync(profilePath)) {
    warning(`Shell profile not found: ${profilePath}`);
    return;
  }

  let content: string;
  try {
    content = fs.readFileSync(profilePath, "utf-8");
  } catch {
    warning(`Could not read shell profile: ${profilePath}`);
    return;
  }

  if (content.includes("arc shell-init")) {
    check(`Shell integration found in ${profilePath}`);
  } else {
    warning(`Shell integration not found in ${path.basename(profilePath)}`);
  }
}

function formatExpiry(expiresAt: number): string {
  const now = Date.now();
  const diff = expiresAt - now;
  const absDiff = Math.abs(diff);

  const minutes = Math.floor(absDiff / 60000);
  const hours = Math.floor(minutes / 60);
  const days = Math.floor(hours / 24);

  let timeStr: string;
  if (days > 0) {
    timeStr = `${days}d ${hours % 24}h`;
  } else if (hours > 0) {
    timeStr = `${hours}h ${minutes % 60}m`;
  } else {
    timeStr = `${minutes}m`;
  }

  return diff > 0 ? `expires in ${timeStr}` : `expired ${timeStr} ago`;
}

async function checkProfiles(): Promise<void> {
  let config;
  try {
    config = loadConfig();
  } catch {
    // Config already reported as broken above
    return;
  }

  const names = Object.keys(config.profiles);
  if (names.length === 0) {
    console.log(`  ${pc.dim("No profiles configured.")}`);
    return;
  }

  console.log();
  console.log("  " + pc.bold("Profiles:"));

  for (const name of names) {
    const profile = config.profiles[name];
    const tool = profile.tool ?? "claude";
    const isActive = name === config.activeProfile;
    const label = isActive ? pc.bold(name) + pc.dim(" *") : name;
    console.log(`    ${label} ${pc.dim(`(${tool} / ${profile.authType})`)}`);

    // Binary check
    if (findBinary(tool)) {
      console.log(`      ${pc.green("\u2714")} ${tool} binary found`);
    } else {
      console.log(`      ${pc.red("\u2716")} ${tool} binary not found`);
    }

    // Credential check
    try {
      const cred = await getCredentialStatus(profile);
      if (cred.authenticated) {
        let authMsg = "";
        switch (cred.method) {
          case "oauth":
            authMsg = "Authenticated";
            if (cred.expiresAt !== undefined) {
              authMsg += ` (${formatExpiry(cred.expiresAt)})`;
            }
            break;
          case "api-key":
            authMsg = "API key configured";
            break;
          case "env-var":
            authMsg = "API key set via env var";
            break;
          case "bedrock":
            authMsg = "Bedrock credentials configured";
            break;
          case "vertex":
            authMsg = "Vertex credentials configured";
            break;
          case "foundry":
            authMsg = "Foundry credentials configured";
            break;
        }
        console.log(`      ${pc.green("\u2714")} ${authMsg}`);
      } else {
        if (cred.expired) {
          const expiryInfo = cred.expiresAt !== undefined ? ` (${formatExpiry(cred.expiresAt)})` : "";
          console.log(`      ${pc.yellow("\u26A0")} Credentials expired${expiryInfo}`);
        } else {
          console.log(`      ${pc.red("\u2716")} Not authenticated`);
        }
      }
    } catch {
      console.log(`      ${pc.red("\u2716")} Error checking credentials`);
    }
  }
}

// ── Main Handler ────────────────────────────────────

export async function handleDoctor(): Promise<void> {
  console.log();
  sectionHeader("ARC Doctor");
  console.log();

  checkConfigFile();
  checkPath();
  checkShellIntegration();

  await checkProfiles();

  console.log();
}

File: rust/src/commands/doctor.rs
use anyhow::Result;
use colored::Colorize;
use std::fs;
use std::path::Path;

use crate::auth::get_credential_status;
use crate::commands::launch::find_binary;
use crate::config::{load_config};
use crate::display::section_header;
use crate::paths::get_config_path;

fn check(msg: &str) {
    println!("  {} {}", "✔".green(), msg);
}

fn cross(msg: &str) {
    println!("  {} {}", "✖".red(), msg);
}

fn warning(msg: &str) {
    println!("  {} {}", "⚠".yellow(), msg);
}

fn check_config_file() -> bool {
    let config_path = get_config_path();

    if !config_path.exists() {
        cross(&format!("Config file not found at {}", config_path.display()));
        return false;
    }

    let raw = match fs::read_to_string(&config_path) {
        Ok(r) => r,
        Err(_) => {
            cross(&format!("Config file unreadable at {}", config_path.display()));
            return false;
        }
    };

    match serde_json::from_str::<crate::types::ArcConfig>(&raw) {
        Ok(cfg) if cfg.version == 1 => {
            check("Config file valid");
            true
        }
        _ => {
            cross("Config file contains malformed JSON or invalid structure");
            false
        }
    }
}

fn check_path() {
    let local_bin = dirs::home_dir()
        .unwrap_or_default()
        .join(".local")
        .join("bin");

    let path_env = std::env::var("PATH").unwrap_or_default();
    let separator = if cfg!(windows) { ";" } else { ":" };

    let on_path = path_env.split(separator).any(|d| {
        Path::new(d)
            .canonicalize()
            .ok()
            .zip(local_bin.canonicalize().ok())
            .map(|(a, b)| a == b)
            .unwrap_or_else(|| d == local_bin.to_string_lossy().as_ref())
    });

    if on_path {
        check("PATH includes ~/.local/bin");
    } else {
        warning("~/.local/bin is not on PATH");
    }
}

fn detect_shell_type() -> &'static str {
    if cfg!(windows) {
        return "powershell";
    }
    let shell_env = std::env::var("SHELL").unwrap_or_default();
    let base = Path::new(&shell_env)
        .file_name()
        .and_then(|n| n.to_str())
        .unwrap_or("");
    match base {
        "zsh" => "zsh",
        "fish" => "fish",
        _ => "bash",
    }
}

fn get_shell_profile_path(shell: &str) -> Option<std::path::PathBuf> {
    let home = dirs::home_dir()?;
    match shell {
        "powershell" => Some(home.join("Documents").join("PowerShell").join("Microsoft.PowerShell_profile.ps1")),
        "bash" => Some(home.join(".bashrc")),
        "zsh" => Some(home.join(".zshrc")),
        "fish" => Some(home.join(".config").join("fish").join("config.fish")),
        _ => None,
    }
}

fn check_shell_integration() {
    let shell = detect_shell_type();
    let profile_path = match get_shell_profile_path(shell) {
        Some(p) => p,
        None => {
            warning(&format!(
                "Could not determine shell profile path for {}",
                shell
            ));
            return;
        }
    };

    if !profile_path.exists() {
        warning(&format!("Shell profile not found: {}", profile_path.display()));
        return;
    }

    let content = match fs::read_to_string(&profile_path) {
        Ok(c) => c,
        Err(_) => {
            warning(&format!(
                "Could not read shell profile: {}",
                profile_path.display()
            ));
            return;
        }
    };

    if content.contains("arc shell-init") {
        check(&format!("Shell integration found in {}", profile_path.display()));
    } else {
        warning(&format!(
            "Shell integration not found in {}",
            profile_path.file_name().unwrap_or_default().to_string_lossy()
        ));
    }
}

fn format_expiry(expires_at_ms: i64) -> String {
    use std::time::{SystemTime, UNIX_EPOCH};
    let now = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_millis() as i64)
        .unwrap_or(0);

    let diff = expires_at_ms - now;
    let abs_diff = diff.unsigned_abs() as u64;
    let minutes = abs_diff / 60000;
    let hours = minutes / 60;
    let days = hours / 24;

    let time_str = if days > 0 {
        format!("{}d {}h", days, hours % 24)
    } else if hours > 0 {
        format!("{}h {}m", hours, minutes % 60)
    } else {
        format!("{}m", minutes)
    };

    if diff > 0 {
        format!("expires in {}", time_str)
    } else {
        format!("expired {} ago", time_str)
    }
}

fn check_profiles() {
    let config = match load_config() {
        Ok(c) => c,
        Err(_) => return,
    };

    let names: Vec<String> = config.profiles.keys().cloned().collect();
    if names.is_empty() {
        println!("  {}", "No profiles configured.".dimmed());
        return;
    }

    println!();
    println!("  {}", "Profiles:".bold());

    for name in &names {
        let profile = &config.profiles[name];
        let tool = profile.tool.as_deref().unwrap_or("claude");
        let is_active = name == &config.active_profile;
        let label = if is_active {
            format!("{}{}", name.bold(), " *".dimmed())
        } else {
            name.to_string()
        };
        println!(
            "    {} {}",
            label,
            format!("({} / {})", tool, profile.auth_type).dimmed()
        );

        // Binary check
        if find_binary(tool) {
            println!("      {} {} binary found", "✔".green(), tool);
        } else {
            println!("      {} {} binary not found", "✖".red(), tool);
        }

        // Credential check
        match get_credential_status(profile) {
            Ok(cred) => {
                if cred.authenticated {
                    let auth_msg = match cred.method.as_str() {
                        "oauth" => {
                            let expiry_info = cred
                                .expires_at
                                .map(|e| format!(" ({})", format_expiry(e)))
                                .unwrap_or_default();
                            format!("Authenticated{}", expiry_info)
                        }
                        "api-key" => "API key configured".to_string(),
                        "env-var" => "API key set via env var".to_string(),
                        "bedrock" => "Bedrock credentials configured".to_string(),
                        "vertex" => "Vertex credentials configured".to_string(),
                        "foundry" => "Foundry credentials configured".to_string(),
                        other => format!("Authenticated ({})", other),
                    };
                    println!("      {} {}", "✔".green(), auth_msg);
                } else if cred.expired == Some(true) {
                    let expiry_info = cred
                        .expires_at
                        .map(|e| format!(" ({})", format_expiry(e)))
                        .unwrap_or_default();
                    println!(
                        "      {} Credentials expired{}",
                        "⚠".yellow(),
                        expiry_info
                    );
                } else {
                    println!("      {} Not authenticated", "✖".red());
                }
            }
            Err(_) => {
                println!("      {} Error checking credentials", "✖".red());
            }
        }
    }
}

pub fn handle_doctor() -> Result<()> {
    println!();
    section_header("ARC Doctor");
    println!();

    check_config_file();
    check_path();
    check_shell_integration();
    check_profiles();

    println!();
    Ok(())
}
